doc/developer/design/20230705_v2_txn_management.md

```diff
@@ -0,0 +1,264 @@
+    "write-only single-table txn".
+
+The rest of this document concerns the implementation of the durability
+component. Additional detail on the timestamp selection is left for a separate
+design document.
+
+## Detailed description
+
+WIP initially just sketching these out while we develop an opinion on which one
+to go with
+
+General notes:
+- There's a fundamental tension between retries/contention and availability
+  here. At the end of the day, if there's a write txn in some other process
+  that's trying to commit at a given timestamp and it overlaps with what you're
+  doing, you either have to force it to retry or wait for it. (Which latter
+  might never complete b/c distributed systems.) As a result, we play tricks
+  with minimizing the cost of retries, tighter contention surface areas, and
+  pushing retries off of important and/or costly txns.
+
+### Option 1: Put all tables in a single persist shard
+
+- Model each update in the shard as something like `(GlobalId, Row)`
+- Reads would then pull out the table(s) they are interested in, likely
+  optimized in some way by pushdown
+- Pro: Can be implemented entirely on top of persist.
+- Pro: Atomic multi-table writes are trivial, as is advancing the upper of all
+  tables at once.
+- Con: Unintuitive. Nothing else works this way.
+- Con: There are likely to be "gotchas" from putting small+large, low+high
+  throughput tables all in one shard. E.g. when individual retractions from some
+  table consolidate out would be dependent on all the other tables
+
+### Option 2: Put all tables in a single persist "state"
+
+- Currently, persist has a 1:1 relationship between shards and the metadata
+  (batches, capabilities, etc) kept by the state machine. Instead, make the
+  metadata a map of `ShardId -> ShardMetadata`.
+- Pro: Atomic multi-table writes are trivial, as is advancing the upper of all
+  tables at once.
+- Con: Greatly adds complexity to the internals of persist, which is already
+  complex.
+- Con: Greatly complicates the API of persist.
+- Con: Membership changes (the mapping between a shard and which state it lives
+  in) are tricky.
+- Con: Significant risk of destabilizing persist, which has to work perfectly
+  for basically anything in Materialize to work.
+- Con: Reading or writing a single table would require loading the metadata of
+  all co-located shards, which scales arbitrarily large with the number of
+  tables. This likely implies an upper bound on the number of tables we can
+  support.
+- Con: Also likely includes a rewrite of how persist monitoring and debugging
+  tools work.
+
+### Option 3: Table remap shard
+
+- A single _table remap_ shard is introduced per environment, which maps _data
+  shard_ uppers to realtime.
+- Intuition: A data shard's upper is no longer the source of truth for when you
+  can read it. Instead, the remap shard is. This allows us to store uncommitted
+  data in the data shards.
```

Comment:
Any early thoughts on how this will get plumbed into the read path? Will `persist_source` somehow abstract this all away? Will it be a layer on top of `persist_source`?

Reply:
with Bens (in-flight) fast-persist-peeks idea, that logic would also have to run in those peeker tasks though, right?